applications/TrilinosApplication/trilinos_space.h

```diff
@@ -418,26 +420,31 @@ class TrilinosSpace
         const bool EnforceInitialGraph = false
         )
     {
+        // Define first auxiliary matrix
+        std::vector<int> NumNz;
+        MatrixType aux_1(::Copy, rA.RowMap(), NumNz.data());
+
+        // First multiplication
+        TransposeMult(rB, rD, aux_1, {true, false}, CallFillCompleteOnResult, KeepAllHardZeros);
+
         // If we enforce the initial connectivity
         if (EnforceInitialGraph) {
-            // Define first auxiliary matrix
-            MatrixType aux(::Copy, rA.Graph());
+            // Second multiplication
+            MatrixType aux_2(::Copy, rA.RowMap(), NumNz.data());
+            Mult(aux_1, rB, aux_2, CallFillCompleteOnResult, KeepAllHardZeros);
 
-            // First multiplication
-            TransposeMult(rB, rD, aux, {true, false}, CallFillCompleteOnResult, true);
+            // Create an Epetra_Matrix
+            MatrixType* aux_3 =  new MatrixType(::Copy, CombineMatricesGraphs(rA, aux_2));
 
-            // Empty the solution Epetra_Matrix
-            SetToZero(rA);
+            // Copy values
+            CopyMatrixValues(*aux_3, aux_2);
```

Comment:
The non-zeros are not the same, so if you try to add a value to the matrix you may get an error. If you define in the multiplication to keep the hard-non-zeros it will raise an error due to an error in Epetra, that the developers (I contacted them) don't plan to fix as Epetra is going to be gone in 2 years, and they recommend switching to Tpetra. This is the fix I could find after several days of debugging and testing. Everything related with sparse matrix multiplication in Epetra is very buggy (that are not going to be fixed), and I assure you this is the best I can do.